common/transport/amqp/src/receiver_link.ts

```diff
@@ -80,98 +149,236 @@ export class ReceiverLink  extends EventEmitter implements AmqpLink {
+            manageReceiverHandlers('on');
+          },
+          receiverOpenEvent: (context: EventContext) => {
+            debug('In receiver attaching state - open event for ' + context.receiver.name);
+            let callback = this._attachingCallback;
+            this._attachingCallback = null;
+            this._fsm.transition('attached', callback);
+          },
+          receiverErrorEvent: (context: EventContext) => {
+            debug('In receiver attaching state - error event for ' + context.receiver.name + ' error is: ' + this._getErrorName(context.receiver.error));
+            this._indicatedError = context.receiver.error;
+            //
+            // We don't transition at this point in that we are guaranteed that the error will be followed by a receiver_close
+            // event.
+            //
+          },
+          receiverCloseEvent: (context: EventContext) => {
+            debug('in receiver attaching state - close event for ' + context.receiver.name);
+            //
+            // We enabled the event listeners on the onEnter handler.  They are to stay alive until we
+            // are about to transition to the detached state.
+            // We are about to transition to the detached state, so clean up.
+            //
+            manageReceiverHandlers('removeListener');
+            let error = this._indicatedError;
+            let callback = this._attachingCallback;
+            this._indicatedError = undefined;
+            this._attachingCallback = undefined;
+            this._receiverCloseOccurred = true;
+            this._fsm.transition('detached', callback, error);
+          },
+          attach: (null),
+          detach: (callback, err) => {
+            /*Codes_SRS_NODE_AMQP_RECEIVER_LINK_06_002: [If the `detach` method is invoked on the `ReceiverLink` while still attaching, the ReceiverLink shall detach.  It will indicate the error to the callback for the `detach` as well as the callback to the `attach`.] */
+            debug('Detaching while attaching of rhea receiver link ' + this._linkAddress);
+            manageReceiverHandlers('removeListener');
+            //
+            // We may have a callback outstanding on the request that started the attaching.
+            // We will signal to that callback that an error occurred. We will also invoke the callback supplied
+            // for this detach.
+            //
+            let error = err || this._indicatedError || new Error('Unexpected link detach while attaching');
+            let attachingCallback = this._attachingCallback;
+            this._indicatedError = undefined;
+            this._attachingCallback = undefined;
+            if (attachingCallback) {
+              attachingCallback(error);
             }
-
-            this._fsm.transition('detached', undefined, err);
+            this._fsm.transition('detached', callback, error);
```

Comment:
something feels awkward about this.
1. you have to maintain your global state separately of the FSM state which is error prone
2. you're calling 2 callbacks, for attaching and then passing another callback around to detach (which will call that callback with the error you pass down). Are there 2 callers with 2 concurrent operations that each need their own callback?

Reply:
Yes for 2.

---
In reply to: [209360358](https://github.com/Azure/azure-iot-sdk-node/pull/344#discussion_r209360358) [](ancestors = 209360358)